test/docker-e2e/e2e_upgrade_test.go

```diff
@@ -0,0 +1,199 @@
+package docker_e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"celestiaorg/celestia-app/test/docker-e2e/dockerchain"
+
+	"github.com/celestiaorg/celestia-app/v5/pkg/appconsts"
+	"github.com/celestiaorg/tastora/framework/testutil/wait"
+)
+
+// TestCelestiaAppMinorUpgrade tests a simple upgrade from one minor version to another.
+func (s *CelestiaTestSuite) TestCelestiaAppMinorUpgrade() {
+	t := s.T()
+	if testing.Short() {
+		t.Skip("skipping upgrade test in short mode")
+	}
+
+	ctx := context.Background()
+	const (
+		baseVersion   = "v4.0.2-mocha"
+		targetVersion = "v4.0.7-mocha"
+	)
+
+	cfg := dockerchain.DefaultConfig(s.client, s.network)
+	cfg.Tag = baseVersion
+	builder := dockerchain.NewCelestiaChainBuilder(t, cfg)
+
+	chain, err := builder.Build(ctx)
+	s.Require().NoError(err)
+
+	err = chain.Start(ctx)
+	s.Require().NoError(err)
+
+	s.Require().NoError(wait.ForBlocks(ctx, 5, chain))
+
+	err = chain.Stop(ctx)
+	s.Require().NoError(err)
+
+	chain.UpgradeVersion(ctx, targetVersion)
+
+	err = chain.Start(ctx)
+	s.Require().NoError(err)
+
+	err = wait.ForBlocks(ctx, 2, chain)
+	s.Require().NoError(err)
+
+	// Verify the version after upgrade
+	validatorNode := chain.GetNodes()[0]
+
+	rpcClient, err := validatorNode.GetRPCClient()
+	s.Require().NoError(err, "failed to get RPC client for version check")
+
+	abciInfo, err := rpcClient.ABCIInfo(ctx)
+	s.Require().NoError(err, "failed to fetch ABCI info")
+	s.Require().Equal(strings.TrimPrefix(targetVersion, "v"), abciInfo.Response.GetVersion(), "version mismatch")
+}
+
+// TestCelestiaAppMajorUpgrade tests a major upgrade from v4.0.7-mocha to a commit hash which has v5
+// using the signaling mechanism.
+func (s *CelestiaTestSuite) TestCelestiaAppMajorUpgrade() {
+	t := s.T()
+	if testing.Short() {
+		t.Skip("skipping major upgrade test in short mode")
+	}
+
+	ctx := context.Background()
+	const (
+		baseVersion   = "v4.0.7-mocha"
+		targetVersion = "d13d38a"
+		targetAppVer  = uint64(5) // The expected app_version after upgrade
+		chainID       = appconsts.TestChainID
+		homeDir       = "/var/cosmos-chain/celestia"
```

Comment:
these constants as well as the function I described lower down can maybe go into the test suite file, as they may be relevant to more tests, not just this one.